proposals/0000-no-incomplete.rst

```diff
@@ -0,0 +1,268 @@
+  Standardizing ``Incomplete`` could make it harder standardize ``GADTs`` and those other extensions later.
+
+  See the future work section for how this is mitigated.
+
+- This proposal effectively promotes some GHC warnings to a language extension.
+  While GHC strives for comparability, ensuring programs with ``-Werror=...``, once accepted, continue to be accepted in future releases was lower down on the prioritization of aspects of compatibility GHC sought to ensure.
+  This makes the completeness checker improvements changes more risky, as they now entail somewhat stronger comparability guarantees.
+
+  See the future work section for how this is mitigated.
+
+- In the short term, we have far too many knobs to control the same thing with each new ``-fdefer-*`` flag corresponding to an existing warning.
+  But, my hope is that in future versions of the language ``Incomplete`` can be deprecated and then removed, reducing the number of knobs back to something sane.
+
+- Somebody is going to think this has something to do with Kurt Gödel unless we choose a different name.
+
+Possible Future Work
+--------------------
+
+The drawbacks are serious enough that I feel compelled to sketch some future plans that would mitigate them.
+
+Missing vs intentionally omitted patterns
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The chief drawback with pattern matching today as it relates to completeness checking is that the Haskell implementation cannot tell whether a user skipped a pattern because they think it's impossible or because they forgot.
+If we had something vaguely like Agda's "absurd patterns", however, the user could use those to use those cover the impossible the cases, making clear that anything else the user really did forget.
+
+  The following examples use a completely strawman syntax.
+  As this isn't part of the proposal proper I don't wish to take up more space in this proposal working out a proper spec.
+
+Using absurd patterns for types like plain ``Void`` may seem low value::
+
+  \case {}
+
+Is hardly worse than::
+
+  \case ABSURD
+
+Laziness also helps, my making one have to force nested patterns anyways.
+But therein we also get a new shorthand::
+
+  \case Identity (v :: Void) -> case v of {}
+
+can become::
+
+  \case Identity !ABSURD
+
+
+Finally, the real benefit is with absurdities stemming from type equalities.
+For example, given:
+
+::
+
+  data G :: Type -> Type where
+    GBool :: G Bool
+    GInt :: G Int
+
+instead of:
+
+::
+
+  f :: G Bool -> ...
+  f GBool -> ...
+
+do something like:
+
+::
+
+  f :: G Bool -> ...
+  f GBool -> ...
+  f (ABSURD @ GInt)
+
+Firstly, this should help make code more self-documenting and should allow for better error/warning messages.
+But more relevant to the problem at hand, this could possibly allow simpler Haskell implementations that aren't so sophisticated that they can derive proofs of pattern impossibility very well on their own, but can verify user-written arguments.
```

Comment:
Sorry, I still don't get this. How does `ABSURD` help simplify the coverage checking problem?

Presumably
```hs
f (ABSURD @ GInt)
```
is (imaginary syntax for) a programmer assertion that the `GInt` case is impossible. Is the idea that this assertion will be unchecked, and itself constitutes the programmer's explicitly specified reason for the program failing at runtime, should it in fact turn out to be invalid? Or if it is checked, why is that easier than treating the omission of the constructor as such an assertion?

I can imagine extending Core/System FC with a language of checkable "apartness evidence" that would turn the coverage checker into an elaborator. That would allow a simpler lint algorithm to verify the absence of incomplete matches in Core. But I highly doubt the evidence terms would be anything a programmer would want to read or write, and that doesn't seem to be what you are proposing in the short term (though it comes up in the following section)?

Reply:
> Or if it is checked, why is that easier than treating the omission of the constructor as such an assertion?

It is meant to be checked. `(ABSURD @ GInt)` is a simple example, but imagine if this was buried between deep within a more complicated pattern. The compiler can just follow the given pattern, rather than try to come up with it.

I will admit perhaps it's more saving the compiler work doing in exhaustive search than allowing it to be less sophisticated, as this doesn't show how to combine the equalities to get the absurdity. Still perhaps it could be augmented with something along those lines (type applications in patterns helps a bit) that gets us closer.